backends/simplego/dotgeneral_block.go

```diff
@@ -0,0 +1,323 @@
+	crossSize       int
+	contractingSize int
+
+	// Axes from the original tensor shape (needed for copying flat → blocked)
+	contractingAxes []int
+	batchAxes       []int
+}
+
+// Equal implements nodeDataComparable for de-duplication.
+func (d *blockForDotGeneralData) Equal(other nodeDataComparable) bool {
+	o, ok := other.(*blockForDotGeneralData)
+	if !ok {
+		return false
+	}
+	return d.blockLog2Dim == o.blockLog2Dim &&
+		d.blockedShape.Equal(o.blockedShape) &&
+		d.batchSize == o.batchSize &&
+		d.crossSize == o.crossSize &&
+		d.contractingSize == o.contractingSize &&
+		slices.Equal(d.contractingAxes, o.contractingAxes) &&
+		slices.Equal(d.batchAxes, o.batchAxes)
+}
+
+// Compile-time check that blockForDotGeneralData implements nodeDataComparable.
+var _ nodeDataComparable = (*blockForDotGeneralData)(nil)
+
+func init() {
+	setNodeExecutor(backends.OpTypeBlockForDotGeneral, priorityGeneric, execBlockForDotGeneral)
+}
+
+// shouldPreBlock determines if a tensor should be pre-blocked for DotGeneral.
+// Unlike shouldPreBlockRHS, this works for both LHS and RHS with any normalized shape.
+//
+// Pre-blocking is beneficial when:
+// - The tensor is a constant or parameter (not computed dynamically)
+// - The tensor is large enough to benefit from blocking
+// - The dtype is supported for blocking
+//
+// Parameters:
+//   - node: the input node to potentially block
+//   - crossSize, contractingSize: normalized sizes for blocking calculation
+func shouldPreBlock(node *Node, crossSize, contractingSize int) bool {
+	dtype := node.shape.DType
+
+	// Check dtype is supported for blocking
+	switch dtype {
+	case dtypes.Float32, dtypes.Float64,
+		dtypes.Float16, dtypes.BFloat16,
+		dtypes.Int8, dtypes.Uint8,
+		dtypes.Int16, dtypes.Int32, dtypes.Int64,
+		dtypes.Uint16, dtypes.Uint32, dtypes.Uint64:
+		// dtype supported, continue to size check
+	default:
+		return false
+	}
+
+	// Must be large enough to benefit from blocking
+	blockDim := 1 << DotGeneralTargetBlockLog2Dim[dtype]
+	if crossSize < blockDim || contractingSize < blockDim {
+		return false
+	}
+
+	// Only pre-block constants or parameters (not computed values)
+	// This ensures we only pay the blocking cost once
+	switch node.opType {
+	case backends.OpTypeParameter, backends.OpTypeConstant:
+		return true
+	default:
+		return false
+	}
+}
+
+// shouldPreBlockRHS determines if the RHS (weights) should be pre-blocked for DotGeneral.
+// This is a specialized version for backward compatibility with 2D weight matrices.
+//
+// Pre-blocking is beneficial when:
+// - RHS is a 2D tensor [K, N] (standard weight layout)
+// - RHS has no batch dimensions (weights are shared across batch)
+// - Single contracting axis on axis 0 for RHS
+// - The dtype is supported for blocking
+// - The matrix is large enough to benefit from blocking (at least one full block in each dimension)
+func shouldPreBlockRHS(rhs *Node, lhsContractingAxes, rhsContractingAxes, rhsBatchAxes []int) bool {
```

Comment:
For DotGeneral we don't want to specialize RHS or LHS.  Maybe for the "MatMul" path we want. 

Maybe call `shouldPreBlockMatMulRHS` ?

Reply:
Good call - I remove the `shouldPreBlockRHS` and `blockRHSForDotGeneral` from this PR. 

This PR now just provides the general-purpose `blockForDotGeneral()` infrastructure, and #266 (SmallMatMul) will add the logic with proper `MatMul`-specific naming